fs/move.py

```diff
@@ -62,17 +66,42 @@ def move_file(
                 _src_fs.move(
                     src_path, dst_path, overwrite=True, preserve_time=preserve_time
                 )
-            else:
-                # Standard copy and delete
-                with _src_fs.lock(), _dst_fs.lock():
-                    copy_file(
-                        _src_fs,
-                        src_path,
-                        _dst_fs,
-                        dst_path,
-                        preserve_time=preserve_time,
-                    )
-                    _src_fs.remove(src_path)
+                return
+
+            if _src_fs.hassyspath(src_path) and _dst_fs.hassyspath(dst_path):
+                # if both filesystems have a syspath we create a new OSFS from a
+                # common parent folder and use it to move the file.
+
+                # we have to raise ResourceReadOnly manually if a FS is read-only
+                if _src_fs.getmeta().get("read_only", True):
+                    raise ResourceReadOnly(_src_fs, src_path)
+                if _dst_fs.getmeta().get("read_only", True):
+                    raise ResourceReadOnly(_dst_fs, dst_path)
+
+                try:
+                    src_syspath = _src_fs.getsyspath(src_path)
+                    dst_syspath = _dst_fs.getsyspath(dst_path)
+                    common = commonpath([src_syspath, dst_syspath])
+                    rel_src = frombase(common, src_syspath)
+                    rel_dst = frombase(common, dst_syspath)
+                    with _src_fs.lock(), _dst_fs.lock(), open_fs(
+                        common, writeable=True
+                    ) as base:
+                        base.move(rel_src, rel_dst, preserve_time=preserve_time)
+                    return  # optimization worked, exit early
+                except ValueError:
+                    pass
```

Comment:
Urgh, isn't this generally regarded as a code-smell?

Reply:
Then I'd need two more ifs (one for each `frombase` call) which would result in more nesting, I thought it was cleaner this way

Reply:
At the very least you should add a comment in the code explaining why you're doing this?

Reply:
👍 Did that